functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+
+### Transactions
+
+Each of the operations in our low-level API act on a single entity. However, nearly all databases have a mechanism for allowing users to perform multiple operations as a single _transaction_. (TODO: Reference other chapters on transactional semantics here.) This means: 
+
+* The batch of operations is viewed as a single atomic operation, meaning that either all of the operations succeed together or fail together
+* The database is in a valid state before, and after the transaction
+* The batch update appears to be _isolated_; other queries should never see a database state in which only some of the operations have been applied
+
+We can fulfill these requirements through an interface that consumes a database and a set of operations to be performed, and which produces a database whose state reflects the given changes. All of the changes submitted in the batch should be applied through the addition of a _single_ layer. However, we have one small problem: All of the functions we wrote in our low-level API add a new layer to the database. If we were to perform a batch with _n_ operations, we would thus see _n_ new layers added, when what we would really like is to have exactly 1 new layer.   
+
+The key insight here is that the layer we want is the _top_ layer that would be produced by performing those updates in sequence. Therefore, the solution is to execute each of the user’s operations one after another, each of which will create a new layer. When the last layer is created, we take only that top layer and place it on the initial database (leaving all the intermediate layers to pine for the fjords). Only after we've done all this will we update the database's timestamp.
+All this is done in the *transact-on-db* function, that receives the initial value of the database and the batch of operations to perform, and returns its updated value. 
+
+````clojure
+(defn transact-on-db [initial-db ops]
+    (loop [[op & rst-ops] ops transacted initial-db]
+      (if op
+          (recur rst-ops (apply (first op) transacted (rest op)))
+          (let [initial-layer  (:layers initial-db)
+                new-layer (last (:layers transacted))]
+            (assoc initial-db :layers (conj  initial-layer new-layer) :curr-time (next-ts initial-db) :top-id (:top-id transacted))))))
+```` 
+Note here that we used the term _value_, this means that only the caller to this function is exposed to the updated state, and all other users of the database are unaware of this change (as a database is a value, and therefore cannot change). 
+In order to have a system where users can be exposed to state changes performed by others, users do not interact directly with the database, but rather refer to it using another level of indirection [REF HERE]. This additional level is implemented using Clojure's element called *Atom*, which is a one of Clojure's reference types. Here we leverage two key features of an *Atom*, which are:

```

Comment:
You mention two key features and then list three :)

I found this section hard to understand. I had to go away and read up on Clojure's *Atom* to figure it out. I think this was partly because you use the term *element* here in a different way to how you use it elsewhere in the chapter. It's also the first use of *dereference* and I get the feeling that it has a special meaning in Clojure-land. I think the 3rd point you list is the relevant one and points 1 and 2 muddy the explanation for me.

It seems that the important thing is that you pass an atom (which references the thing you want to "update") and a function (which does the updating) to the transaction, and you get back a *new* thing which is the result of the function being applied to the atom. Do I have that right?